### PR TITLE
PEP 688: Making the buffer protocol accessible in Python

### PR DESCRIPTION
**Pull request by JelleZijlstra**

### CODE REVIEW

File: pep-0688.rst [hunk 1]
```diff
@@ -0,0 +1,252 @@
+PEP: 688
+Title: Making the buffer protocol accessible in Python
+Author: Jelle Zijlstra <jelle.zijlstra@gmail.com>
+Sponsor: Jelle Zijlstra <jelle.zijlstra@gmail.com>
```
Comment (AA-Turner) [inline pep-0688.rst:4]: Core developers don't need sponsors
```suggestion
```

File: pep-0688.rst [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 23-Apr-2022
+Python-Version: 3.12
+
+
+Abstract
+========
+
+This PEP proposes a mechanism to inspect in Python whether a type implements
```
Comment (AA-Turner) [inline pep-0688.rst:15]: Phrasing suggestion
```suggestion
This PEP proposes a mechanism within Python to inspect if a type implements
```

File: pep-0688.rst [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+objects that implement the protocol.
+
+
+Motivation
+==========
+
+The CPython C API provides a versatile mechanism for accessing the
+underlying memory of an object, the buffer protocol from :pep:`3118`.
+Functions that accept binary data are usually written to accept any
+object implementing the buffer protocol. For example, as I write this,
```
Comment (AA-Turner) [inline pep-0688.rst:26]: Phrasing suggestion
```suggestion
object implementing the buffer protocol. For example, at the time of writing,
```

File: pep-0688.rst [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+Type checkers may provide additional pointers in their error messages. For example,
+when they encounter a place where a buffer object is passed to a function that
+is annotated to only accept ``bytes``, the error message could include a note suggesting
+to use ``types.Buffer`` instead.
+
+
+Reference Implementation
+========================
+
+[Link to any existing implementation and details about its state, e.g. proof-of-concept.]
```
Comment (AA-Turner) [inline pep-0688.rst:214]: Perhaps remove the entire section until there's a reference implementation?
```suggestion
```

File: pep-0688.rst [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+options would be useful in the type system. For example, typeshed
+currently provides separate type aliases for writable and read-only
+buffers.
+
+However, in the C buffer protocol, these options cannot be
+queried directly on the type object. The only way to figure out
+whether an object supports a writable buffer is to actually
+ask for the buffer. For some types, such as ``memoryview``,
+whether the buffer is writable depends on the instance:
+some instances are read-only and others are not. As such, I propose to
```
Comment (AA-Turner) [inline pep-0688.rst:90]: Phrasing suggestion
```suggestion
some instances are read-only and others are not. As such, we propose to
```

**Review by AA-Turner — COMMENTED** [review]
There are two uses of "I" which I've suggested turning into plural or neutral language as the rest of the document seems to use "we".

A couple of small suggestions elsewise.

A

File: pep-0688.rst (hunk 4 above)
Comment (JelleZijlstra) [reply pep-0688.rst:214]: But there is a reference implementation (after a lot of segfaults): https://github.com/python/cpython/compare/main...JelleZijlstra:typesbuffer?expand=1. I'll add a link to it.

Comment (AA-Turner) [reply pep-0688.rst:214]: Ah, sorry! I'm only slowly getting half back into things, so I may have missed discussion on this in other fora.

A

Comment (JelleZijlstra) [reply pep-0688.rst:214]: No, you were right! I only just pushed that implementation. Thanks for the review, I'm applying your suggestions right now.

File: pep-0688.rst (hunk 2 above)
Comment (JelleZijlstra) [reply pep-0688.rst:15]: Thanks, but I prefer the current wording here

Comment (CAM-Gerlach) [reply pep-0688.rst:14]: For what its worth, I also find the current phrasing somewhat hard to parse and understand, particularly so when it is the very first sentence of the PEP's abstract and thus there isn't much in the way of context clues to easily infer the intended interpretation. What about these other options that also address the issue?

```suggestion
This PEP proposes a mechanism for Python code to inspect whether a type implements
```

```suggestion
This PEP proposes a Python-level mechanism to inspect whether a type implements
```

```suggestion
This PEP proposes a mechanism to inspect at the Python level whether a type implements
```

```suggestion
This PEP proposes a mechanism in Python to inspect whether a type implements
```

**Comment by CAM-Gerlach** [comment]
Hey @JelleZijlstra , would you like me to proofread and suggest copyedits on this?

**Comment by JelleZijlstra** [comment]
@CAM-Gerlach yes please! Can't guarantee I'll agree with all your suggestions though :)

File: pep-0688.rst (hunk 2 above)
Comment (JelleZijlstra) [reply pep-0688.rst]: Thanks, I like the first suggestion here. Just pushed that change.

File: pep-0688.rst [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+The ``types.Buffer`` class does not require any special treatment
+in type checkers.
+
+Equivalent for older Python versions
+------------------------------------
+
+New typing features are usually backported to older Python versions
+in the ``typing_extensions`` package. Because the buffer protocol
+is accessible only in C, ``types.Buffer`` cannot be implemented
+in a pure Python package. As a temporary workaround, a
```
Comment (CAM-Gerlach) [inline pep-0688.rst:154]:
```suggestion
in a pure-Python package like ``typing_extensions``.
As a temporary workaround, a
```

Add missing hyphen, and make the connection with ``typing_extensions`` more explicit (it wasn't immediately obvious to me at first, since its not actually stated currently)

File: pep-0688.rst [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+``typing_extensions.Buffer`` ABC will be provided on Python versions
```
Comment (CAM-Gerlach) [inline pep-0688.rst:155]:
```suggestion
``typing_extensions.Buffer`` `Abstract Base Class <Buffer ABC_>`__
will be provided on Python versions
```

Despite being used a number of times, what "ABC" stands for is never actually mentioned in this PEP, and it would seem to make sense here to link to the section describing this particular ABC in more detail.

File: pep-0688.rst [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+
+No special meaning for ``bytes``
+--------------------------------
+
+The special case stating that ``bytes`` may be used as a shorthand
+for other ``ByteString`` types will be removed from the ``typing``
+documentation.
+With ``types.Buffer`` available as an alternative, there is no good
+reason to allow ``bytes`` as a shorthand.
+We suggest that type checkers that implement this behavior should deprecate and
```
Comment (CAM-Gerlach) [inline pep-0688.rst:173]: Avoid double "that" and ensure it is clear that you're referring to the current behavior, not implementing new behavior specified in this PEP.

```suggestion
We suggest that type checkers currently implementing this behavior should deprecate and
```

or

```suggestion
We suggest that type checkers which currently implement this behavior should deprecate and
```

File: pep-0688.rst [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+
+Backwards Compatibility
+=======================
+
+As the runtime changes in this PEP only add a new class, there are
+no backwards compatibility concerns.
+
+However, the recommendation to remove the special behavior for
+``bytes`` in type checkers does have backwards compatibility
+impact on users. An `experiment <https://github.com/python/mypy/pull/12661>`__
```
Comment (CAM-Gerlach) [inline pep-0688.rst:185]:
```suggestion
``bytes`` in type checkers does have a backwards compatibility
impact on their users. An `experiment <https://github.com/python/mypy/pull/12661>`__
```

Better grammar and be more explicit that the backwards compatibility impact is only on _type checker_ users, not Python as a language.

File: pep-0688.rst [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+with mypy shows that several major open source projects type
+checked with mypy will see new errors if the ``bytes`` promotion
```
Comment (CAM-Gerlach) [inline pep-0688.rst:187]:
```suggestion
with mypy shows that several major open source projects that use it
for type checking will see new errors if the ``bytes`` promotion
```

Avoid awkward repetition and improve readibility

File: pep-0688.rst [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+is removed. However, many of these errors can be fixed by improving
+the stubs in typeshed, as already done for the
+`builtins <https://github.com/python/typeshed/pull/7631>`__,
+`binascii <https://github.com/python/typeshed/pull/7677>`__,
+`pickle <https://github.com/python/typeshed/pull/7678>`__ modules.
```
Comment (CAM-Gerlach) [inline pep-0688.rst:192]:
```suggestion
and `pickle <https://github.com/python/typeshed/pull/7678>`__ modules.
```

Fix grammar error

File: pep-0688.rst [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+Overall, the change improves overall type safety,
```
Comment (CAM-Gerlach) [inline pep-0688.rst:193]:
```suggestion
Overall, the change improves type safety,
```

Avoid double "overall", "Overall the change overall".

There are a few other options to fix this, if you don't care for this one

File: pep-0688.rst [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+
+
+How to Teach This
+=================
+
+We will add notes pointing to ``types.Buffer`` to appropriate places in the
+documentation, such as `typing.readthedocs.io <https://typing.readthedocs.io/en/latest/>`__
+and the `mypy cheat sheet <https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html>`__.
+Type checkers may provide additional pointers in their error messages. For example,
+when they encounter a place where a buffer object is passed to a function that
```
Comment (CAM-Gerlach) [inline pep-0688.rst:210]:
```suggestion
when they encounter a buffer object being passed to a function that
```

Removing unnecessary awkward wording to simplify and clarify phrasing

File: pep-0688.rst [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+is annotated to only accept ``bytes``, the error message could include a note suggesting
+to use ``types.Buffer`` instead.
```
Comment (CAM-Gerlach) [inline pep-0688.rst:212]:
```suggestion
the use of ``types.Buffer`` instead.
```

Cleaner phrasing and avoid grammatically awkward  "suggesting to use"

File: pep-0688.rst [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+
+
+Rejected Ideas
+==============
+
+Buffer ABC
+----------
+
+An `earlier proposal <https://github.com/python/cpython/issues/71688>`__ suggested
+adding a ``collections.abc.Buffer`` ABC to represent buffer objects. This idea
```
Comment (CAM-Gerlach) [inline pep-0688.rst:230]:
```suggestion
adding a ``collections.abc.Buffer`` `abstract base class <https://docs.python.org/3/glossary.html#term-abstract-base-class>`__
to represent buffer objects. This idea
```

As before, might be good actually mention once what "ABC" means and link to it, in case any readers are unfamiliar

File: pep-0688.rst [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+stalled because an ABC with no methods does not fit well into the ``collections.abc``
+module. Furthermore, it required manual registration of buffer classes, including
+those in the standard library. This PEP's approach of using the ``__instancecheck__``
+hook is more natural and does not require explicit registration.
+Nevertheless, the ABC proposal has the advantage that it does not require C changes,
+and we are proposing to adopt a version of it in the third-party ``typing_extensions``
+package for the benefit of users of older versions of Python.
```
Comment (CAM-Gerlach) [inline pep-0688.rst:237]:
```suggestion
package for the benefit of users on older Python versions.
```

Simplify the syntax a bit to make it easier to read and parse, and avoid triple "of"

File: pep-0688.rst [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+
+Read-only and writable buffers
+------------------------------
+
+To avoid making changes to the buffer protocol itself, this PEP currently
+does not provide a way to distinguish between read-only and writable buffers.
+That's unfortunate, because some APIs require a writable buffer, and one of
+the most common buffer types (``bytes``) is always read-only.
+Should we add a new mechanism in C to declare that a type implementing the
+buffer protocol is always read-only?
```
Comment (CAM-Gerlach) [inline pep-0688.rst:251]: Sidenote: At least at the level of static type checking, wouldn't the opposite make more sense, i.e. declaring that a type is writable, since a writable buffer is a subtype of a read-only buffer? This also follows the general principle that for the base collections, etc. e.g. `Set`, `Mapping`, `Sequence`, etc, the base type is read-only, with mutable subtypes. But I'm not a typing expert like you, and there may be other C-implementation reasons this framing makes more sense.

**Review by CAM-Gerlach — COMMENTED** [review]
Here you go! In a lot of places, I provided multiple different possible options for you to consider that resolved an issue, if you didn't like the first one (or I myself had no strong preference).

File: pep-0688.rst [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+
+types.Buffer
+------------
+
+A new class ``types.Buffer`` will be added. It cannot be instantiated or
+subclassed, but supports the ``__instancecheck__`` and
+``__subclasscheck__`` hooks.  In CPython, these will check for the presence of the
+``bf_getbuffer`` slot in the type object:
+
+.. code-block:: pycon
```
Comment (Fidget-Spinner) [inline pep-0688.rst:105]: Freudian slip :)?
```suggestion
.. code-block:: python
```

Comment (CAM-Gerlach) [reply pep-0688.rst:105]: Umm, [no](https://pygments.org/docs/lexers/#pygments.lexers.python.PythonConsoleLexer)...

Comment (JelleZijlstra) [reply pep-0688.rst:105]: It confuses me every time I see it too :)

Comment (Fidget-Spinner) [reply pep-0688.rst:105]: Ahh Python console. Cool! Didn't know that. Sorry!

File: pep-0688.rst [hunk 19]
```diff
@@ -0,0 +1,258 @@
...
+Created: 23-Apr-2022
+Python-Version: 3.12
+
+
+Abstract
+========
+
+This PEP proposes a mechanism for Python code to inspect whether a
+type implements the C-level buffer protocol. This allows type
+checkers to check for objects that implement the protocol.
```
[the remark this answers is not in the document]
Comment (JelleZijlstra) [reply pep-0688.rst:16]: Thanks, I'll go with the first one. I don't think the second sentence is necessary.

I'm applying the changes locally and will push them when I'm done, in case I want to make more substantive edits.

File: pep-0688.rst [hunk 20]
```diff
@@ -0,0 +1,258 @@
...
+
+
+Motivation
+==========
+
+The CPython C API provides a versatile mechanism for accessing the
+underlying memory of an object, the buffer protocol from :pep:`3118`.
+Functions that accept binary data are usually written to accept any
+object implementing the buffer protocol. For example, at the time of writing,
```
[the remark this answers is not in the document]
Comment (JelleZijlstra) [reply pep-0688.rst:25]: I considered that, but we're still months away from 3.11 formally existing, and the number may change between whatever is on main now and the 3.11.0 release.

File: pep-0688.rst [hunk 21]
```diff
@@ -0,0 +1,258 @@
...
+there are about 130 functions in CPython using the Argument Clinic
+``Py_buffer`` type, which accepts the buffer protocol.
+
+Currently, there is no way to inspect in Python whether an object
+implements the buffer protocol. Relatedly, the static type system
```
[the remark this answers is not in the document]
Comment (JelleZijlstra) [reply pep-0688.rst:30]: Let's go with "Moreover"

File: pep-0688.rst [hunk 22]
```diff
@@ -0,0 +1,258 @@
...
+currently provides separate type aliases for writable and read-only
+buffers.
+
+However, in the C buffer protocol, these options cannot be
+queried directly on the type object. The only way to figure out
+whether an object supports a writable buffer is to actually
+ask for the buffer. For some types, such as ``memoryview``,
+whether the buffer is writable depends on the instance:
+some instances are read-only and others are not. As such, we propose to
+support only whether a type implements the buffer protocol at
```
[the remark this answers is not in the document]
Comment (JelleZijlstra) [reply pep-0688.rst:90]: Your previous suggestion was about the same sentence, adopted the wording from that one instead.

And now I need to go for a few hours; I'll push what I have so far and come back for the rest later. Thanks for the thorough review!

File: pep-0688.rst (hunk 18 above)
Comment (CAM-Gerlach) [reply pep-0688.rst:105]: Yeah, you're certainly not the first person to think it was a mistake—in fact, I'm pretty sure I did myself at some point :)

File: pep-0688.rst (hunk 22 above)
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply pep-0688.rst]: Sorry, I meant to delete that one in favor of this one before I submitted my review, but I forgot. My other one still has the issue of the repetition of "support" and its use to refer to two totally different things in the same sentence, which is potentially confusing, which this one fixes.

File: pep-0688.rst [hunk 23]
```diff
@@ -0,0 +1,258 @@
...
+Content-Type: text/x-rst
+Created: 23-Apr-2022
+Python-Version: 3.12
+
+
+Abstract
+========
+
+This PEP proposes a mechanism for Python code to inspect whether a
+type implements the C-level buffer protocol. This allows type
```
[the remark this answers is not in the document]
Comment (CAM-Gerlach) [reply pep-0688.rst:15]: This was :+1: ed but it was also marked resolved without applying it. Was that intentional?

File: pep-0688.rst (hunk 16 above)
Comment (CAM-Gerlach) [reply pep-0688.rst:238]: Looks like you fixed this already in another commit :+1:

File: pep-0688.rst (hunk 23 above)
[the remark this answers is not in the document]
Comment (JelleZijlstra) [reply pep-0688.rst:15]: Oops, I think I only changed the similar sentence on line 30. Changing this one now too.

File: pep-0688.rst (hunk 22 above)
[the remark this answers is not in the document]
Comment (JelleZijlstra) [reply pep-0688.rst]: Good point, pushing a version of this one.

File: pep-0688.rst [hunk 24]
```diff
@@ -0,0 +1,258 @@
...
+.. code-block:: python
+
+   def need_buffer(b: Buffer) -> memoryview:
+       return memoryview(b)
+
+   need_buffer(b"xy")  # ok
+   need_buffer("xy")  # rejected by static type checkers
+
+Usage in stubs
+--------------
```
[the remark this answers is not in the document]
Comment (JelleZijlstra) [reply pep-0688.rst:130]: Thanks, I'll go with "stub files"

File: pep-0688.rst (hunk 11 above)
Comment (JelleZijlstra) [reply pep-0688.rst]: Fixed when I added another PR to the list

File: pep-0688.rst (hunk 12 above)
Comment (JelleZijlstra) [reply pep-0688.rst:194]: Yes, also added some more words here.

File: pep-0688.rst (hunk 17 above)
Comment (JelleZijlstra) [reply pep-0688.rst:264]: That's a good point actually. At the Python level, I'd want a WritableBuffer subclass of Buffer that bytearray implements but bytes doesn't. I'll reword the question to point in that direction.

File: pep-0688.rst [hunk 25]
```diff
@@ -0,0 +1,275 @@
...
+
+
+Motivation
+==========
+
+The CPython C API provides a versatile mechanism for accessing the
+underlying memory of an object—the buffer protocol from :pep:`3118`.
+Functions that accept binary data are usually written to handle any
+object implementing the buffer protocol. For example, at the time of writing,
+there are about 130 functions in CPython using the Argument Clinic
```
Comment (AlexWaygood) [inline pep-0688.rst:26]:
```suggestion
there are around 130 functions in CPython using the Argument Clinic
```

File: pep-0688.rst [hunk 26]
```diff
@@ -0,0 +1,275 @@
...
+``Py_buffer`` type, which accepts the buffer protocol.
+
+Currently, there is no way for Python code to inspect whether an object
+supports the buffer protocol. Moreover, the static type system
+does not provide a type annotation to represent the protocol.
+This is a `common problem <https://github.com/python/typing/issues/593>`__
+when type-annotating code that accepts generic buffers.
```
Comment (AlexWaygood) [inline pep-0688.rst:33]:
```suggestion
when adding type annotations to code that accepts generic buffers.
```

File: pep-0688.rst [hunk 27]
```diff
@@ -0,0 +1,275 @@
...
+For static typing purposes, types defined in C extensions usually
+require stub files, as :pep:`described in PEP 484 <484#stub-files>`.
+In stub files, ``types.Buffer`` may be used as a base class to
+indicate that a class implements the buffer protocol.
+
+For example, ``bytes`` may be declared as follows in a stub:
+
+.. code-block:: python
+
+   class bytes(types.Buffer, Sequence[int]):
```
Comment (AlexWaygood) [inline pep-0688.rst:141]: Hmm, does this PEP propose deprecating `typing.ByteString`? If not, surely we'd do this in typeshed, since `ByteString` inherits from `Sequence[int]` in `typing.pyi`.

```suggestion
   class bytes(types.Buffer, typing.ByteString):
```

File: pep-0688.rst [hunk 28]
```diff
@@ -0,0 +1,275 @@
...
+       def decode(self, ...): ...
+       ...
+
+The ``types.Buffer`` class does not require any special treatment
+in type checkers.
```
Comment (AlexWaygood) [inline pep-0688.rst:146]:
```suggestion
by type checkers.
```

File: pep-0688.rst [hunk 29]
```diff
@@ -0,0 +1,275 @@
...
+Equivalent for older Python versions
+------------------------------------
+
+New typing features are usually backported to older Python versions
+in the `typing_extensions <https://pypi.org/project/typing-extensions/>`_
+package. Because the buffer protocol
+is accessible only in C, ``types.Buffer`` cannot be implemented
+in a pure-Python package like ``typing_extensions``. As a temporary
+workaround, a ``typing_extensions.Buffer``
+`Abstract Base Class <Buffer ABC_>`__ will be provided on Python versions
```
Comment (AlexWaygood) [inline pep-0688.rst:157]: The capitalisation looks a bit weird to me here. Also "on" -> "for".

```suggestion
`abstract base class <Buffer ABC_>`__ will be provided for Python versions
```

File: pep-0688.rst [hunk 30]
```diff
@@ -0,0 +1,275 @@
...
+that do not have ``types.Buffer`` available. For the benefit of
```
Comment (AlexWaygood) [inline pep-0688.rst:158]:
```suggestion
that do not have ``types.Buffer`` available.

For the benefit of
```

File: pep-0688.rst [hunk 31]
```diff
@@ -0,0 +1,275 @@
...
+static type checkers, ``typing_extensions.Buffer`` can be used as
+a base class in stubs to mark types as supporting the buffer protocol.
+For runtime uses, the ``ABC.register`` API can be used to register
+buffer classes with ``typing_extensions.Buffer``. When
+``types.Buffer`` is available, ``typing_extensions`` should simply
+re-export it.
```
Comment (AlexWaygood) [inline pep-0688.rst:164]: Do we need to state this in the PEP? -- isn't this something of a `typing_extensions` implementation detail?

File: pep-0688.rst [hunk 32]
```diff
@@ -0,0 +1,275 @@
...
+
+
+No special meaning for ``bytes``
+--------------------------------
+
+The special case stating that ``bytes`` may be used as a shorthand
+for other ``ByteString`` types will be removed from the ``typing``
+documentation.
+With ``types.Buffer`` available as an alternative, there is no good
```
Comment (AlexWaygood) [inline pep-0688.rst:173]:
```suggestion
With ``types.Buffer`` available as an alternative, there will be no good
```

File: pep-0688.rst [hunk 33]
```diff
@@ -0,0 +1,275 @@
...
+As the runtime changes in this PEP only add a new class, there are
+no backwards compatibility concerns.
+
+However, the recommendation to remove the special behavior for
+``bytes`` in type checkers does have a backwards compatibility
+impact on their users. An `experiment <https://github.com/python/mypy/pull/12661>`__
+with mypy shows that several major open source projects that use it
+for type checking will see new errors if the ``bytes`` promotion
+is removed. However, many of these errors can be fixed by improving
+the stubs in typeshed, as already done for the
```
Comment (AlexWaygood) [inline pep-0688.rst:191]:
```suggestion
the stubs in typeshed, as has already been done for the
```

File: pep-0688.rst [hunk 34]
```diff
@@ -0,0 +1,275 @@
...
+Overall, the change improves type safety and makes the type system
+more consistent, so we believe the migration cost is worth it.
+
+
+Security Implications
+=====================
+
+None.
+
+
```
Comment (AlexWaygood) [inline pep-0688.rst:205]: This may just be me being a PEP reviewer noob, but: do we need this section at all if there are 0 security implications? Other typing PEPs don't seem to have it.

```suggestion
```

File: pep-0688.rst [hunk 35]
```diff
@@ -0,0 +1,275 @@
...
+How to Teach This
+=================
+
+We will add notes pointing to ``types.Buffer`` to appropriate places in the
```
Comment (AlexWaygood) [inline pep-0688.rst:209]:
```suggestion
We will add notes pointing to ``types.Buffer`` in appropriate places in the
```

File: pep-0688.rst [hunk 36]
```diff
@@ -0,0 +1,275 @@
...
+documentation, such as `typing.readthedocs.io <https://typing.readthedocs.io/en/latest/>`__
+and the `mypy cheat sheet <https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html>`__.
+Type checkers may provide additional pointers in their error messages. For example,
```
Comment (AlexWaygood) [inline pep-0688.rst:212]:
```suggestion
Type checkers may also provide additional pointers in their error messages. For example,
```

File: pep-0688.rst [hunk 37]
```diff
@@ -0,0 +1,275 @@
...
+adding a ``collections.abc.Buffer``
+`abstract base class <https://docs.python.org/3/glossary.html#term-abstract-base-class>`__
+to represent buffer objects. This idea
+stalled because an ABC with no methods does not fit well into the ``collections.abc``
+module. Furthermore, it required manual registration of buffer classes, including
+those in the standard library. This PEP's approach of using the ``__instancecheck__``
+hook is more natural and does not require explicit registration.
+Nevertheless, the ABC proposal has the advantage that it does not require C changes,
+and we are proposing to adopt a version of it in the third-party ``typing_extensions``
+package for the benefit of users of older Python versions.
```
Comment (AlexWaygood) [inline pep-0688.rst:242]: Split into two paragraphs (and one sentence into two sentences)

```suggestion

Nevertheless, the ABC proposal has the advantage that it does not require C changes.
This PEP proposes to adopt a version of it in the third-party ``typing_extensions``
package for the benefit of users of older Python versions.
```

**Review by AlexWaygood — COMMENTED** [review]
I'm excited to see this happen! A few thoughts from me below :)

File: pep-0688.rst (hunk 26 above)
Comment (JelleZijlstra) [reply pep-0688.rst:33]: Going to go with "writing type annotations for code", which I think better describes typeshed

File: pep-0688.rst (hunk 27 above)
Comment (JelleZijlstra) [reply pep-0688.rst:141]: Honestly I'm not really sure what ByteString is good for. I'll use `memoryview` as an example instead.

Comment (AlexWaygood) [reply pep-0688.rst:141]: > Honestly I'm not really sure what ByteString is good for.

I mean... neither...

File: pep-0688.rst (hunk 29 above)
Comment (JelleZijlstra) [reply pep-0688.rst:157]: Agree, https://docs.python.org/3/library/abc.html lowercases it.

File: pep-0688.rst (hunk 31 above)
Comment (JelleZijlstra) [reply pep-0688.rst:164]: This is important if people do `typing_extensions.Buffer.register(MyFancyBuffer)`. I'll put a note to make that explicit.

File: pep-0688.rst (hunk 11 above)
Comment (CAM-Gerlach) [reply pep-0688.rst]: Yep, noticed that—I'd thought I'd left a comment and marked this as resolved, but it must not have gone through for some reason.

File: pep-0688.rst (hunk 36 above)
Comment (JelleZijlstra) [reply pep-0688.rst:212]: Feel like that's redundant with "additional"

Comment (AlexWaygood) [reply pep-0688.rst:212]: Yeah fair, feel free to ignore this one then!

File: pep-0688.rst (hunk 29 above)
Comment (CAM-Gerlach) [reply pep-0688.rst]: I initially lowercased it, but I evidently uppercased it to signify its use later as an initialism. But its not terribly important either way.

File: pep-0688.rst (hunk 27 above)
Comment (JelleZijlstra) [reply pep-0688.rst]: I guess it means "bytes or bytearray", which is a fairly useful concept.

File: pep-0688.rst (hunk 34 above)
Comment (CAM-Gerlach) [reply pep-0688.rst]: I'd think so, since it actually explicitly states such, rather than leaving the reader to hopefully assume that by the fact that it wasn't addressed.

File: pep-0688.rst [hunk 38]
```diff
@@ -0,0 +1,274 @@
...
+writable buffers.
+
+Specification
+=============
+
+types.Buffer
+------------
+
+A new class, ``types.Buffer``, will be added. It cannot be instantiated or
+subclassed, but supports the ``__instancecheck__`` and
```
Comment (AlexWaygood) [inline pep-0688.rst:101]: We can subclass it in stubs, as discussed below

```suggestion
subclassed at runtime, but supports the ``__instancecheck__`` and
```

File: pep-0688.rst (hunk 35 above)
Comment (CAM-Gerlach) [reply pep-0688.rst]: This doesn't fully solve the problem, since now "in" is repeated twice rather than "to", which is still potentially confusing and hard to read. It seems like I can't make a suggestion here, but what about something like 

```rst
We will add notes pointing to ``types.Buffer`` in appropriate places throughout the
```

File: pep-0688.rst [hunk 39]
```diff
@@ -0,0 +1,274 @@
...
+in a pure-Python package like ``typing_extensions``. As a temporary
+workaround, a ``typing_extensions.Buffer``
+`abstract base class <Buffer ABC_>`__ will be provided for Python versions
+that do not have ``types.Buffer`` available.
+
+For the benefit of
+static type checkers, ``typing_extensions.Buffer`` can be used as
+a base class in stubs to mark types as supporting the buffer protocol.
+For runtime uses, the ``ABC.register`` API can be used to register
+buffer classes with ``typing_extensions.Buffer``. When
```
Comment (AlexWaygood) [inline pep-0688.rst:163]:
```suggestion
buffer classes with ``typing_extensions.Buffer``.

When
```

File: pep-0688.rst [hunk 40]
```diff
@@ -0,0 +1,274 @@
...
+
+As the runtime changes in this PEP only add a new class, there are
+no backwards compatibility concerns.
+
+However, the recommendation to remove the special behavior for
+``bytes`` in type checkers does have a backwards compatibility
+impact on their users. An `experiment <https://github.com/python/mypy/pull/12661>`__
+with mypy shows that several major open source projects that use it
+for type checking will see new errors if the ``bytes`` promotion
+is removed. However, many of these errors can be fixed by improving
```
Comment (AlexWaygood) [inline pep-0688.rst:194]: "However" is used two sentences back

```suggestion
is removed. Nonetheless, many of these errors can be fixed by improving
```

**Review by AlexWaygood — COMMENTED** [review]
A few more small points :)

File: pep-0688.rst (hunk 40 above)
Comment (JelleZijlstra) [reply pep-0688.rst]: I just removed "However"

File: pep-0688.rst (hunk 39 above)
Comment (JelleZijlstra) [reply pep-0688.rst]: More paragraph breaks, I like it

**Review by AlexWaygood — APPROVED** [review]
Looks great to me :)

**Comment by JelleZijlstra** [comment]
Thanks everyone for the feedback! I just merged the PR so it shows up on the official site. If you have further textual suggestions, feel free to put a note here, open a PR, give me a handwritten note at PyCon, or let me know in whatever other way you can think of.

**Comment by gvanrossum** [comment]
Carrier pigeon.

**Comment by CAM-Gerlach** [comment]
Interpretive dance.
